File: swift-bridge/Sources/AVSpeechSynthesizerBridge/Provider.swift
import AVFAudio
import Foundation

struct AVSProviderVoiceConfigPayload: Codable {
    var name: String
    var identifier: String
    var primaryLanguages: [String]
    var supportedLanguages: [String]
}

final class AVSProviderVoiceBox: NSObject {
    let voice: AVSpeechSynthesisProviderVoice

    init(voice: AVSpeechSynthesisProviderVoice) {
        self.voice = voice
        super.init()
    }
}

final class AVSProviderRequestBox: NSObject {
    let request: AVSpeechSynthesisProviderRequest

    init(request: AVSpeechSynthesisProviderRequest) {
        self.request = request
        super.init()
    }
}

private func avsProviderVoiceBox(_ token: UnsafeMutableRawPointer?) throws -> AVSProviderVoiceBox {
    guard let token else {
        throw AVSBridgeError.invalidArgument("missing provider voice token")
    }
    return avsBorrow(token)
}

private func avsProviderRequestBox(_ token: UnsafeMutableRawPointer?) throws -> AVSProviderRequestBox {
    guard let token else {
        throw AVSBridgeError.invalidArgument("missing provider request token")
    }
    return avsBorrow(token)
}

private func avsProviderVoicePayload(from voice: AVSpeechSynthesisProviderVoice) -> AVSProviderVoicePayload {
    AVSProviderVoicePayload(
        name: voice.name,
        identifier: voice.identifier,
        primaryLanguages: voice.primaryLanguages,
        supportedLanguages: voice.supportedLanguages,
        voiceSize: voice.voiceSize,
        version: voice.version,
        gender: Int(voice.gender.rawValue),
        age: Int(voice.age)
    )
}

@_cdecl("avs_provider_voice_new_json")
public func avs_provider_voice_new_json(
    _ configJson: UnsafePointer<CChar>?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    do {
        let config = try avsDecodeJSON(configJson, as: AVSProviderVoiceConfigPayload.self)
        let voice = AVSpeechSynthesisProviderVoice(
            name: config.name,
            identifier: config.identifier,
            primaryLanguages: config.primaryLanguages,
            supportedLanguages: config.supportedLanguages
        )
        return avsRetain(AVSProviderVoiceBox(voice: voice))
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return nil
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return nil
    }
}

@_cdecl("avs_provider_voice_release")
public func avs_provider_voice_release(_ token: UnsafeMutableRawPointer?) {
    guard let token else { return }
    avsRelease(token)
}

@_cdecl("avs_provider_voice_snapshot_json")
public func avs_provider_voice_snapshot_json(
    _ token: UnsafeMutableRawPointer?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    do {
        let voice = try avsProviderVoiceBox(token).voice
        return avsCString(try avsEncodeJSON(avsProviderVoicePayload(from: voice)))
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return nil
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return nil
    }
}

@_cdecl("avs_provider_voice_set_voice_size")
public func avs_provider_voice_set_voice_size(
    _ token: UnsafeMutableRawPointer?,
    _ voiceSize: Int64,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        let voice = try avsProviderVoiceBox(token).voice
        voice.voiceSize = voiceSize
        return AVS_OK
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return error.statusCode
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return AVS_UNKNOWN
    }
}

@_cdecl("avs_provider_voice_set_version")
public func avs_provider_voice_set_version(
    _ token: UnsafeMutableRawPointer?,
    _ version: UnsafePointer<CChar>?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        let voice = try avsProviderVoiceBox(token).voice
        voice.version = try avsRequireString(version, field: "version")
        return AVS_OK
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return error.statusCode
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return AVS_UNKNOWN
    }
}

@_cdecl("avs_provider_voice_set_gender")
public func avs_provider_voice_set_gender(
    _ token: UnsafeMutableRawPointer?,
    _ gender: Int64,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        guard let rawGender = AVSpeechSynthesisVoiceGender(rawValue: Int(gender)) else {
            throw AVSBridgeError.invalidArgument("unknown provider voice gender: \(gender)")
        }
        let voice = try avsProviderVoiceBox(token).voice
        voice.gender = rawGender
        return AVS_OK
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return error.statusCode
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return AVS_UNKNOWN
    }
}

@_cdecl("avs_provider_voice_set_age")
public func avs_provider_voice_set_age(
    _ token: UnsafeMutableRawPointer?,
    _ age: Int64,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    do {
        let voice = try avsProviderVoiceBox(token).voice
        voice.age = Int(age)
        return AVS_OK
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return error.statusCode
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return AVS_UNKNOWN
    }
}

@_cdecl("avs_provider_voice_update_speech_voices")
public func avs_provider_voice_update_speech_voices(
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    _ = outErrorMessage
    AVSpeechSynthesisProviderVoice.updateSpeechVoices()
    return AVS_OK
}

@_cdecl("avs_provider_request_new")
public func avs_provider_request_new(
    _ voiceToken: UnsafeMutableRawPointer?,
    _ ssmlRepresentation: UnsafePointer<CChar>?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    do {
        let voice = try avsProviderVoiceBox(voiceToken).voice
        let ssmlRepresentation = try avsRequireString(
            ssmlRepresentation,
            field: "provider request SSML representation"
        )
        let request = AVSpeechSynthesisProviderRequest(
            ssmlRepresentation: ssmlRepresentation,
            voice: voice
        )
        return avsRetain(AVSProviderRequestBox(request: request))
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return nil
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return nil
    }
}

@_cdecl("avs_provider_request_release")
public func avs_provider_request_release(_ token: UnsafeMutableRawPointer?) {
    guard let token else { return }
    avsRelease(token)
}

@_cdecl("avs_provider_request_ssml_representation")
public func avs_provider_request_ssml_representation(
    _ token: UnsafeMutableRawPointer?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    do {
        let request = try avsProviderRequestBox(token).request
        return avsCString(request.ssmlRepresentation)
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return nil
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return nil
    }
}

@_cdecl("avs_provider_request_copy_voice")
public func avs_provider_request_copy_voice(
    _ token: UnsafeMutableRawPointer?,
    _ outErrorMessage: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    do {
        let request = try avsProviderRequestBox(token).request
        return avsRetain(AVSProviderVoiceBox(voice: request.voice))
    } catch let error as AVSBridgeError {
        outErrorMessage?.pointee = avsCString(error.description)
        return nil
    } catch {
        outErrorMessage?.pointee = avsCString(error.localizedDescription)
        return nil
    }
}

File: src/provider.rs
use core::{
    ffi::{c_char, c_void},
    ptr,
};

use serde::{Deserialize, Serialize};

use crate::error::AvSpeechError;
use crate::ffi;
use crate::private::{
    error_from_status, json_cstring, parse_json_ptr, result_from_status, string_from_ptr,
    to_cstring,
};
use crate::voice::SpeechSynthesisVoiceGender;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ProviderVoiceConfig {
    name: String,
    identifier: String,
    primary_languages: Vec<String>,
    supported_languages: Vec<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ProviderVoiceSnapshot {
    name: String,
    identifier: String,
    primary_languages: Vec<String>,
    supported_languages: Vec<String>,
    voice_size: i64,
    version: String,
    gender: i64,
    age: i64,
}

/// Wraps an AVSpeechSynthesis provider voice.
pub struct SpeechSynthesisProviderVoice {
    token: *mut c_void,
}

impl Drop for SpeechSynthesisProviderVoice {
    fn drop(&mut self) {
        if !self.token.is_null() {
            unsafe {
                ffi::provider::avs_provider_voice_release(self.token);
            }
            self.token = ptr::null_mut();
        }
    }
}

impl SpeechSynthesisProviderVoice {
    /// Creates an AVSpeechSynthesis provider voice wrapper.
    pub fn new(
        name: impl Into<String>,
        identifier: impl Into<String>,
        primary_languages: impl IntoIterator<Item = impl Into<String>>,
        supported_languages: impl IntoIterator<Item = impl Into<String>>,
    ) -> Result<Self, AvSpeechError> {
        let config = ProviderVoiceConfig {
            name: name.into(),
            identifier: identifier.into(),
            primary_languages: primary_languages.into_iter().map(Into::into).collect(),
            supported_languages: supported_languages.into_iter().map(Into::into).collect(),
        };
        let config_json = json_cstring(&config)?;
        let mut err_msg: *mut c_char = ptr::null_mut();
        let token = unsafe {
            ffi::provider::avs_provider_voice_new_json(config_json.as_ptr(), &mut err_msg)
        };
        if token.is_null() {
            return Err(unsafe { error_from_status(ffi::status::UNKNOWN, err_msg) });
        }
        Ok(Self { token })
    }

    /// Triggers the AVSpeechSynthesis provider voice catalog to refresh.
    pub fn update_speech_voices() -> Result<(), AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let status =
            unsafe { ffi::provider::avs_provider_voice_update_speech_voices(&mut err_msg) };
        unsafe { result_from_status(status, err_msg) }
    }

    fn snapshot(&self) -> Result<ProviderVoiceSnapshot, AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let json =
            unsafe { ffi::provider::avs_provider_voice_snapshot_json(self.token, &mut err_msg) };
        if !err_msg.is_null() {
            return Err(unsafe { error_from_status(ffi::status::UNKNOWN, err_msg) });
        }
        unsafe { parse_json_ptr(json, "provider voice snapshot") }
    }

    /// Returns the AVSpeechSynthesis provider voice display name.
    pub fn name(&self) -> Result<String, AvSpeechError> {
        Ok(self.snapshot()?.name)
    }

    /// Returns the AVSpeechSynthesis provider voice identifier.
    pub fn identifier(&self) -> Result<String, AvSpeechError> {
        Ok(self.snapshot()?.identifier)
    }

    /// Returns the primary AVSpeechSynthesis languages for this provider voice.
    pub fn primary_languages(&self) -> Result<Vec<String>, AvSpeechError> {
        Ok(self.snapshot()?.primary_languages)
    }

    /// Returns every AVSpeechSynthesis language supported by this provider voice.
    pub fn supported_languages(&self) -> Result<Vec<String>, AvSpeechError> {
        Ok(self.snapshot()?.supported_languages)
    }

    /// Returns the AVSpeechSynthesis provider voice size in bytes.
    pub fn voice_size(&self) -> Result<i64, AvSpeechError> {
        Ok(self.snapshot()?.voice_size)
    }

    /// Returns the AVSpeechSynthesis provider voice version string.
    pub fn version(&self) -> Result<String, AvSpeechError> {
        Ok(self.snapshot()?.version)
    }

    /// Returns the AVSpeechSynthesis provider voice gender.
    pub fn gender(&self) -> Result<SpeechSynthesisVoiceGender, AvSpeechError> {
        Ok(SpeechSynthesisVoiceGender::from_raw(
            self.snapshot()?.gender,
        ))
    }

    /// Returns the AVSpeechSynthesis provider voice age hint.
    pub fn age(&self) -> Result<i64, AvSpeechError> {
        Ok(self.snapshot()?.age)
    }

    /// Sets the AVSpeechSynthesis provider voice size in bytes.
    pub fn set_voice_size(&self, voice_size: i64) -> Result<(), AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let status = unsafe {
            ffi::provider::avs_provider_voice_set_voice_size(self.token, voice_size, &mut err_msg)
        };
        unsafe { result_from_status(status, err_msg) }
    }

    /// Sets the AVSpeechSynthesis provider voice version string.
    pub fn set_version(&self, version: &str) -> Result<(), AvSpeechError> {
        let version = to_cstring(version)?;
        let mut err_msg: *mut c_char = ptr::null_mut();
        let status = unsafe {
            ffi::provider::avs_provider_voice_set_version(
                self.token,
                version.as_ptr(),
                &mut err_msg,
            )
        };
        unsafe { result_from_status(status, err_msg) }
    }

    /// Sets the AVSpeechSynthesis provider voice gender.
    pub fn set_gender(&self, gender: SpeechSynthesisVoiceGender) -> Result<(), AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let status = unsafe {
            ffi::provider::avs_provider_voice_set_gender(self.token, gender.as_raw(), &mut err_msg)
        };
        unsafe { result_from_status(status, err_msg) }
    }

    /// Sets the AVSpeechSynthesis provider voice age hint.
    pub fn set_age(&self, age: i64) -> Result<(), AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let status =
            unsafe { ffi::provider::avs_provider_voice_set_age(self.token, age, &mut err_msg) };
        unsafe { result_from_status(status, err_msg) }
    }
}

/// Wraps an AVSpeechSynthesis provider request.
pub struct SpeechSynthesisProviderRequest {
    token: *mut c_void,
}

impl Drop for SpeechSynthesisProviderRequest {
    fn drop(&mut self) {
        if !self.token.is_null() {
            unsafe {
                ffi::provider::avs_provider_request_release(self.token);
            }
            self.token = ptr::null_mut();
        }
    }
}

impl SpeechSynthesisProviderRequest {
    /// Creates an AVSpeechSynthesis provider request from SSML and a voice.
    pub fn new(
        ssml_representation: &str,
        voice: &SpeechSynthesisProviderVoice,
    ) -> Result<Self, AvSpeechError> {
        let ssml_representation = to_cstring(ssml_representation)?;
        let mut err_msg: *mut c_char = ptr::null_mut();
        let token = unsafe {
            ffi::provider::avs_provider_request_new(
                voice.token,
                ssml_representation.as_ptr(),
                &mut err_msg,
            )
        };
        if token.is_null() {
            return Err(unsafe { error_from_status(ffi::status::UNKNOWN, err_msg) });
        }
        Ok(Self { token })
    }

    /// Returns the AVSpeechSynthesis provider request SSML representation.
    pub fn ssml_representation(&self) -> Result<String, AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let string = unsafe {
            ffi::provider::avs_provider_request_ssml_representation(self.token, &mut err_msg)
        };
        if !err_msg.is_null() {
            return Err(unsafe { error_from_status(ffi::status::UNKNOWN, err_msg) });
        }
        unsafe { string_from_ptr(string, "provider request ssml representation") }
    }

    /// Returns the AVSpeechSynthesis provider voice attached to this request.
    pub fn voice(&self) -> Result<SpeechSynthesisProviderVoice, AvSpeechError> {
        let mut err_msg: *mut c_char = ptr::null_mut();
        let token =
            unsafe { ffi::provider::avs_provider_request_copy_voice(self.token, &mut err_msg) };
        if token.is_null() {
            return Err(unsafe { error_from_status(ffi::status::UNKNOWN, err_msg) });
        }
        Ok(SpeechSynthesisProviderVoice { token })
    }
}
